ext/TensorKitChainRulesCoreExt/constructors.jl

```diff
@@ -17,8 +17,10 @@ function ChainRulesCore.rrule(::typeof(Base.copy), t::AbstractTensorMap)
     return copy(t), copy_pullback
 end
 
+# this rule does not work for generic symmetries, as we currently have no way to
+# project back onto the symmetric subspace
```

Comment:
What do you mean by that? The TensorMap constructor does try to project back, right, i.e. I thought that `TensorMap(convert(Array, t), space(t)) == t` should work for the group-based symmetries. Maybe we need to test this somewhere.

Reply:
Now I see, I guess you mean that you need a `ProjectTo` that acts on the array data and would remove the entries where there needs to be zeros for the `TensorMap` constructor.

Reply:
Exactly. I added some tests for these methods and ran into issues, precisely because of that.
In principle, these rules were mostly defined in the first place to facilitate some testing, so I thought it might be safest to just restrict this to the case where we know it will work, because it wasn't entirely clear to me how to implement this projection for non-abelian symmetries. It's not too hard to project out the entries that have to be zero, but I am not entirely sure how to project to the nearest linear combination of clebsch gordans?
I would also have to think a bit longer to consider if we need to support this projection, or throw an error if the deviation is too large, as there really should not be any contribution to the gradient along the non-symmetric directions

Reply:
Well, for cotangents I think directions that are out of the subspace can just be ignored without warning or error. In the factorisations, for example, I warn for directions that are dual to gauge-dependent directions (e.g. changing the phase of an eigenvector), but I simply ignore cotangent directions that cannot contribute (e.g. changing the norm of an eigenvector).

Reply:
I guess that in principle this error could be made optional:
https://github.com/Jutho/TensorKit.jl/blob/master/src/tensors/tensor.jl#L380

Reply:
I just used `tol=Inf`, I'm running the tests locally right now, I'll push an update in a bit